File: Grind75/55-BinaryTreeRightView/python/BinaryTreeRightView.py
#!/bin/python3

from __future__ import annotations
from collections import deque
import sys
sys.path.append('../../Utility/python')
from typing import Deque, Optional, List, Tuple
from Tree import *

class Solution:
  def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
    if root is None:
      return []
    rightView: List[int] = []
    nodeQ: Deque[Tuple[Optional[TreeNode], int]] = deque()
    nodeQ.append( (root, 0) )
    while nodeQ:
      node, depth = nodeQ.popleft()
      if node is None:
        continue
      if depth == len(rightView):
        rightView.append(node.val)
      else:
        rightView[depth] = node.val
      nodeQ.append( (node.left,  depth + 1) )
      nodeQ.append( (node.right, depth + 1) )
    return rightView

  def rightSideView2(self, root: Optional[TreeNode]) -> List[int]:
    if root is None:
      return []
    rightView: List[int] = []
    levelOrder: List[List[int]] = []
    nodeQ: Deque[Tuple[Optional[TreeNode], int]] = deque()
    nodeQ.append( (root, 0) )
    while nodeQ:
      node, depth = nodeQ.popleft()
      if node is None:
        continue
      if depth == len(levelOrder):
        levelOrder.append([])
      levelOrder[depth].append(node.val)
      nodeQ.append( (node.left,  depth + 1) )
      nodeQ.append( (node.right, depth + 1) )
    for level in levelOrder:
      rightView.append(level[-1])
    return rightView


if __name__ == '__main__':
  soln = Solution()
  tree = makeTree([
      TreeNode(0),
      TreeNode(1), TreeNode(2),
      TreeNode(3), None, TreeNode(4), None,
      None, None, None, None, TreeNode(5),
    ]
  )

  printTree(tree)
  print(soln.rightSideView(tree))

File: Grind75/55-BinaryTreeRightView/cpp/BinaryTreeRightView.cpp
#include "BinaryTreeRightView.h"
#include <queue>
#include <vector>
#include <Tree.h>

using namespace std;
using namespace BST;

vector<int> Solution::rightSideView(TreeNode* root) {
  vector<int> rightView;
  if (!root)
    return rightView;
  queue<TreeNode*> nodeQ;
  nodeQ.push(root);

  while (!nodeQ.empty()) {
    size_t qSize = nodeQ.size();
    TreeNode *rightMost = nullptr;
    for (size_t i = 0; i < qSize; i++) {
      TreeNode *node = nodeQ.front();
      nodeQ.pop();
      if (!node)
        continue;
      rightMost = node;
      nodeQ.push(node->left);
      nodeQ.push(node->right);
    }
    if (rightMost)
      rightView.push_back(rightMost->val);
  }

  return rightView;
}

vector<int> Solution::rightSideView2(TreeNode* root) {
  vector<int> rightView;
  if (!root)
    return rightView;
  queue<pair<TreeNode*, size_t>> nodeQ;
  nodeQ.push( {root, 0} );

  while (!nodeQ.empty()) {
    auto [node,depth] = nodeQ.front();
    nodeQ.pop();
    if (!node)
      continue;
    if (depth == rightView.size())
      rightView.push_back(node->val);
    else
      rightView[depth] = node->val;
    nodeQ.push( {node->left,  depth+1} );
    nodeQ.push( {node->right, depth+1} );
  }

  return rightView;
}

